src/Fhir.Anonymizer.Shared.Core/Utility/SubstituteUtility.cs

```diff
@@ -0,0 +1,126 @@
+                var targetChildren = replacementNode.Children(name).Cast<ElementNode>().ToList();
+
+                int i = 0;
+                foreach (var child in children)
+                {
+                    if (visitedNodes.Contains(child))
+                    {
+                        // Skip replacement if child already processed before.
+                        i++;
+                        continue;
+                    }
+                    else if (i < targetChildren.Count)
+                    {
+                        // We still have target nodes, do replacement
+                        processResult.Update(SubstituteNode(child, targetChildren[i ++], visitedNodes, keepNodes));
+                    }
+                    else if (keepNodes.Contains(child))
+                    {
+                        // Substitute with an empty node when no target node available but we need to keep this node
+                        processResult.Update(SubstituteNode(child, GetDummyNode(), visitedNodes, keepNodes));
+                    }
+                    else
+                    {
+                        // Remove source node when no target node availabe and we don't need to keep the source node
+                        node.Remove(child);
+                        processResult.AddProcessRecord(AnonymizationOperations.Substitute, child);
+                    }
+                }
+
+                while (i < targetChildren.Count)
+                {
+                    // Add extra target nodes, create a new copy before adding
+                    node.Add(s_provider, ElementNode.FromElement(targetChildren[i ++]));
+                }
+            }
+
+            // children nodes not presented in replacement value, we need either remove or keep a dummy copy
+            var nonReplacementChildren = node.Children()
+                .Where(element => !replaceChildrenNames.Contains(element.Name))
+                .Cast<ElementNode>().ToList();
+            foreach (var child in nonReplacementChildren)
+            {
+                if (visitedNodes.Contains(child))
+                {
+                    continue;
+                }
+                else if (keepNodes.Contains(child))
+                {
+                    SubstituteNode(child, GetDummyNode(), visitedNodes, keepNodes);
+                }
+                else
+                {
+                    node.Remove(child);
+                }
+            }
+
+            node.Value = replacementNode.Value;
+            processResult.AddProcessRecord(AnonymizationOperations.Substitute, node);
+            return processResult;
+        }
+
+        // To keep consistent anonymization changes made by preceding rules, we should figure out whether a node can be removed during substitution
+        public static bool ShouldKeepNodeDuringSubstitution(ElementNode node, HashSet<ElementNode> visitedNodes, HashSet<ElementNode> keepNodes)
```

Comment:
Maybe we should rename here? it more like generate a keepNodes record.

Reply:
Thanks. Updated